waku/v2/peermanager/peer_manager.go

```diff
@@ -310,19 +306,15 @@ func (pm *PeerManager) SelectPeer(proto protocol.ID, specificPeers []peer.ID, lo
 	//  - which topics they track
 	//  - latency?
 
+	//Try to fetch from serviceSlot
+	if peerID, err := pm.serviceSlots.getPeers(proto).getRandom(); err == nil {
```

Comment:
First we should not try to fetch from serviceSlot.
Because relay peers are not added to serviceSlots. So, first we should check if proto is relay then return a random peer from relayPeers. If it is not relay, then proceed with service slot selection.

Reply:
But why to do a map lookup unnecessarily which is costly, rather a check for relay would be less costlier. 
Hence i was suggesting the other way around.

Reply:
These are 2 flows i am looking up for selectPeer 
1. with relay -> where a random peer has to be selected from peerStore
2. for service-protocols -> lookup in service-slot to get a peer randomly, if none found then lookup in peer-table(this case should almost never happen). 

If we just check for relayProto first and proceed in case of relay, it would just be a string compare. 
Not sure why FilterPeersByProto would come into picture here. 
And for servicePeer selection, the overhead is just 1 string-compare. 

Now that i look at the code further, in case of relay a call to `getPeers` would end up creating a `serviceSlot` for relay which is not what we want. 
Also, the cost of acquiring a lock on serviceSlot is un-necessary for relay. Instead of all this overhead shouldn't it be easier to check for relay first which is a stringCompare. And if proto is not relay, then lookup in service slot?

Also, i don't think it is a good idea to insert when a GetMethod is called on serviceSlots.

Reply:
This is the previous excerpt of code. 
https://github.com/waku-org/go-waku/blob/master/waku/v2/utils/peer.go#L46-L55. In the SupportedProtocols, we filter protocols for the peer that are needed. Which is filterProtocol*noOfPeers.
<img width="613" alt="image" src="https://github.com/waku-org/go-waku/assets/115948546/29640cd6-c630-4179-86c1-6d4b1c7d2034">
Sorry , I not able to find the single StringCompare that you are talking about.

getPeers create a serviceSlot, though that slot will not have peers for DefaultRelay. I can return nil from getPeers for DefaultRelay but that will external check.

Reply:
OK, i see what you are getting at. 
I was referring to line "proto == WakuRelayIDv200" but we do this check after filtering for peers. 
But now since we are looking into serviceSlots first , for servicePeers we don't need to do this check.

Let us change this behaviour of not creating the serviceSlot while calling getPeers. Especially if it is going to create one for relay, would lead to confusion. 

@5lliot are you @harsh-98 using a different username? Got confused suddenly :)